pkg/reconciler/apibinding/apibinding_reconcile.go

```diff
@@ -0,0 +1,528 @@
+						"error creating CRD %s: %v",
+						crd.Name,
+						err,
+					)
+
+					// Only return for non already-exists errors
+					return reconcileStatusStop, err // retry
+				}
+
+				// If we got an already-exists error, proceed below to set condition
+			}
+
+			r.deletedCRDTracker.Remove(crd.Name)
+
+			conditions.MarkFalse(
+				apiBinding,
+				apisv1alpha1.CRDReady,
+				apisv1alpha1.WaitingForEstablishedReason,
+				conditionsv1alpha1.ConditionSeverityInfo,
+				"waiting for CRD to be established",
+			)
+
+			needToWaitForRequeue = true
+		} else {
+			// We never set Conversion, so nil it out before comparing
+			existingCRD = existingCRD.DeepCopy()
+			existingCRD.Spec.Conversion = nil
+
+			if !reflect.DeepEqual(existingCRD.Spec, crd.Spec) {
+				klog.Infof("CRD diff: %s", cmp.Diff(existingCRD.Spec, crd.Spec))
+				crd.ResourceVersion = existingCRD.ResourceVersion
+
+				if _, err := r.updateCRD(ctx, crd); err != nil {
+					conditions.MarkFalse(
+						apiBinding,
+						apisv1alpha1.CRDReady,
+						apisv1alpha1.UpdateErrorReason,
+						conditionsv1alpha1.ConditionSeverityWarning,
+						"error updating CRD %s to latest schema: %v",
+						crd.Name,
+						err,
+					)
+					return reconcileStatusStop, err // retry
+				}
+
+				needToWaitForRequeue = true
+			} else {
+				if !apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Established) || apihelpers.IsCRDConditionTrue(existingCRD, apiextensionsv1.Terminating) {
+					needToWaitForRequeue = true
+				}
+			}
+		}
+
+		schemaSpecJSON, err := json.Marshal(schema.Spec)
+		if err != nil {
+			// TODO condition
+			return reconcileStatusStop, err
+		}
+		schemaHash := sha256.New()
+		if _, err := schemaHash.Write(schemaSpecJSON); err != nil {
+			// TODO condition
+			return reconcileStatusStop, err
+		}
+
+		boundResources = append(boundResources, apisv1alpha1.BoundAPIResource{
+			Group:    schema.Spec.Group,
+			Resource: schema.Spec.Names.Plural,
+			Schema: apisv1alpha1.BoundAPIResourceSchema{
+				Name: schema.Name,
+				UID:  string(schema.UID),
+				Hash: fmt.Sprintf("%x", schemaHash.Sum(nil)),
+			},
+			StorageVersions: crd.Status.StoredVersions,
+		})
+	}
+
+	conditions.MarkTrue(apiBinding, apisv1alpha1.APIExportValid)
+
+	if needToWaitForRequeue {
+		return reconcileStatusStop, nil
+	}
+
+	conditions.MarkTrue(apiBinding, apisv1alpha1.CRDReady)
+	apiBinding.Status.Initializers = []string{}
+	apiBinding.Status.Phase = apisv1alpha1.APIBindingPhaseBound
+	apiBinding.Status.BoundAPIExport = &apiBinding.Spec.Reference
+	apiBinding.Status.BoundResources = boundResources
+
+	return reconcileStatusContinue, nil
+}
+
+func (r *workspaceAPIExportReferenceReconciler) unbind(apiBinding *apisv1alpha1.APIBinding) {
+	apiBinding.Status.BoundAPIExport = nil
+	apiBinding.Status.BoundResources = nil
+	// TODO(ncdc): what do we do about CRs that were created from APIBindings? Do we disallow unbinding until they're
+	// all deleted? Do we delete them?
+}
```

Comment:
unbind is something we cannot do. We can have stuck bindings with the referenced export being gone, but we cannot unbind. Unbind = delete of CRs.

Reply:
That'll be an admission change, right?

Reply:
If we can consistently protect this in admission, yes. But in a distributed world we might not be able to. There we have to find a graceful behaviour, e.g. either keep the CRs working, but the rebinding stuck, or at worst make the CRs read-only.

Reply:
done. Removed unbind and now setting APIExportValid to false.